airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/jsonschema/JsonRecordMapper.kt

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.integrations.destination.s3.jsonschema
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.ObjectNode
+
+abstract class JsonRecordMapper<R> {
+    fun mapRecordWithSchema(record: JsonNode?, schema: ObjectNode): R {
```

Comment:
I don't understand the nomenclature. In my mind, a record is a list of fields, and the mapper maps fields one by one. I'd imagine having a JsonFieldMapper that one would implement if they want to a field-by-field mapping. JsonRecordMapper would only need to be inherited if there's a whole-record mapping.
The simplest JsonFieldMapper interface would contains a single mapField method, and each implementer is free to do the switch.
The default mapFieldRecord would be a fall-through to deepclone for most of the field types
I don't really understand the parameterized type either. It looks like we only have a need for Json->json mapping, so let's remove the parameter for now.
I really think the schemaMapper is also overkill. It's only used once per stream to initialize the convertedSchema, which points towards having that called as part of the JsonRecordConverter constructor, and kept private

Reply:
**Parameter:** I agree that the parametrization was premature. I'll take it out now and save it for generalizing for the CDK.

**Record versus field:** A record as a group of fields is different than the individual fields. For example, field mapping might be handled differently depending on the depth of the mapped field, which would depend on the record. This would make sense if this were used to simplify CSV conversion. (All top-level fields get serialized to strings.)

**Schema mapper:** It's used once per stream *per format type or feature*, and any chained record mappers need the result of schema mapping from the previous step as an input, which means keeping it as a first class thing makes sense. However, I agree that it should be private relative to the whole record pipeline once this is finalized. In the CDK, I was imagining something like

```
RecordPipelineStep {
   schemaMapper: SchemaMapper
   recordMapper: RecordMapper
}

RecordPipeline(steps: List<RecordPipelineStep>) {
  List<Schema> cache;
  
  init { /* run each schema step in sequence, populating the cache * / }
  
  getSchema() { /* just return the final schema */ }
  getRecord() {
    /*
        result1 <- mapper(record, cache[0])
        result2 <- mapper(result1, cache[1])
        etc...
        return final result
    */
}

```